src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,153 @@
+        }
+    }
+
     @Override
     public void add(T value) {
+        if (isEmpty()) {
+            addFirst(value);
+        } else {
+            addLast(value);
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index out of bound: " + index);
+        }
+        if (isEmpty()) {
+            addFirst(value);
+        } else if (index == size) {
+            addLast(value);
+        } else if (index == 0) {
+            head.prev = new Node<>(null, value, head);;
+            head = head.prev;
+        } else {
+            Node<T> current = getNodeByIndex(index);
+            Node<T> newNode = new Node<>(current.prev, value, current);
+            current.prev.next = current.prev = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element: list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (!isValidIndex(index)) {
+            throw new IndexOutOfBoundsException("Index " + index + " is not valid");
+        }
+        if (index == size - 1) {
+            return tail.item;
+        }
+        return getNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (!isValidIndex(index)) {
+            throw new IndexOutOfBoundsException("Index " + index + " is not valid");
+        }
+        Node<T> current = getNodeByIndex(index);
+        T oldValue = current.item;
+        current.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index " + index + " is not valid");
+        }
+        Node<T> current = getNodeByIndex(index);
+        if (current.prev == null) {
+            head = current.next;
+            current.next = null;
+        } else if (current.next == null) {
+            tail = current.prev;
+            current.prev.next = null;
+            current.prev = null;
+        } else {
+            current.prev.next = current.next;
+            current.next.prev = current.prev;
+        }
+        size--;
+        return current.item;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        int nodeIndex = getIndexByValue(object);
+        if (nodeIndex == -1) {
+            return false;
+        } else {
+            remove(nodeIndex);
+            return true;
+        }
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private boolean isValidIndex(int index) {
+        return index >= 0 && index < size;
+    }
```

Comment:
Delegate throwing exception to this method. Then you can make public methods shorter and avoid repeatable code
```suggestion
    private void checkIndex(int index) {
        if (index < 0 || index >= size) {
            throw....
        }
    }
```